File: src/2015/2/b/js/solution.js
// http://adventofcode.com/2015/day/2

const { byNumber, readInputData } = require('../../../../util');

const input = readInputData(__dirname);

const inputRows = input.split('\n');

const inputDimensions = inputRows
  .map((row) => row
    .split('x')
    .filter(Boolean)
    .sort(byNumber))
  .filter((row) => row.length);

/** Expects box dimensions to be in ascending order */
function computeRibbon(box) {
  const wrap = box[0] * 2 + box[1] * 2;
  const bow = box[0] * box[1] * box[2];
  return wrap + bow;
}

const squareFeet = inputDimensions.reduce((sqFt, box) => {
  return computeRibbon(box) + sqFt;
}, 0);

console.log('The elves require', squareFeet, 'feet of ribbon');

File: src/2015/2/b/hs/solution.hs
-- http://adventofcode.com/2015/day/2
import Data.List.Split
import Data.List

main :: IO()
main = do
    stdin <- getContents
    let squareFeet = inputToSquareFeet stdin
    putStrLn $ "The elves require " ++ show squareFeet ++ " feet of ribbon"

-- use Int insead of Word for simplicity/performance's sake
-- I used a guard instead, but I feel like that too is wrong
-- seems like there are interesting cases for negative surface area
computeRibbonFeet :: Int -> Int -> Int -> Int
computeRibbonFeet smallDimension medDimension bigDimension
    | smallDimension <= 0 = 0
    | medDimension <= 0 = 0
    | bigDimension <= 0 = 0
    | otherwise =
      let wrap = smallDimension * 2 + medDimension * 2
          bow = smallDimension * medDimension * bigDimension
      in wrap + bow

dimensionsToSurfaceArea :: [Int] -> Int
-- this feels really dumb, yay knowledge gaps
dimensionsToSurfaceArea dims =
  let middle = head . tail
      srtd = sort dims
  in computeRibbonFeet (head srtd) (middle srtd) (last srtd)

-- could blow up if not a valid Int :(
strToInt :: String -> Int
strToInt val = read val :: Int

textDimensionsToDimensions :: String -> [Int]
textDimensionsToDimensions text = map strToInt (splitOn "x" text)

textDimensionsToSurfaceArea :: String -> Int
textDimensionsToSurfaceArea = dimensionsToSurfaceArea . textDimensionsToDimensions

textDimListToSurfaceAreaList :: [String] -> [Int]
textDimListToSurfaceAreaList = map textDimensionsToSurfaceArea 

inputToSquareFeet :: String -> Int
inputToSquareFeet input =
  let surfaceAreaList = textDimListToSurfaceAreaList (lines input)
  in foldl (+) 0 surfaceAreaList
